### PR TITLE
Code for OOP Assignment 1

### PR DESCRIPTION
**Pull request by kyptdoan**
Defined and implemented 3 following classes to the code base:
1. Die: to represent a die with 1 method for the die to be rolled.
2. Game: to represent a game with 3 dice, 2 methods to roll, find sum then report.
3. Testing: to implement debug tool to verify if the dice values and the sum are valid.

### CODE REVIEW

File: CMP1903_A1_2324/Program.cs [hunk 1]
```diff
@@ -6,14 +6,88 @@
...
+            /// erroneous input handling and continuous roll execution.
+            /// </summary>
+            while (true)
+            {
+                Console.WriteLine("Choose a function:");
+                Console.WriteLine("1. Type '1' if you want to roll the dice.");
+                Console.WriteLine("2. Type '2' if you want to quit the game.");
+
+                // check if the user inputs an integer.
+                try
```
Comment (KirbyXIV) [inline CMP1903_A1_2324/Program.cs:43]: Good Exception Handling

File: CMP1903_A1_2324/Program.cs [hunk 2]
```diff
@@ -6,14 +6,88 @@
...
+                    user_input = int.Parse(Console.ReadLine());
+                }
+                catch (FormatException) 
+                {
+                    Console.WriteLine("Please only input '1' or '2'.");
+                    continue;
+                }
+
+                // check if the user inputs '1' or '2' for valid options.
+                if (user_input == 1 || user_input == 2)
```
Comment (KirbyXIV) [inline CMP1903_A1_2324/Program.cs:54]: Further Handles Exception Handling

File: CMP1903_A1_2324/Game.cs
Comment (KirbyXIV) [inline CMP1903_A1_2324/Game.cs:1]: Great XML Commenting on every file.

File: CMP1903_A1_2324/Program.cs [hunk 3]
```diff
@@ -6,14 +6,88 @@
...
+                        // Check if the sum of 3 dice values is valid.
+                        testing.CheckExpectation();
+
+                        // Print out the sum of the 3 dice values.
+                        Console.WriteLine(game.FindSum());
+                    }
+                    // option 2 is to quit the game.
+                    else
+                    {
+                        Console.WriteLine("The game ends.");
```
Comment (KirbyXIV) [inline CMP1903_A1_2324/Program.cs:106]: Allows player to exit game

File: CMP1903_A1_2324/Game.cs [hunk 4]
```diff
@@ -1,21 +1,93 @@
...
+        /// Represent the second die of the game.
+        /// </summary>
+        private readonly Die die2;
 
+        /// <summary>
+        /// Represent the third die of the game.
+        /// </summary>
+        private readonly Die die3;
+
+        /// <summary>
```
Comment (KirbyXIV) [inline CMP1903_A1_2324/Game.cs:33]: XML comments for the dies are a bit excessive

File: CMP1903_A1_2324/Testing.cs [hunk 5]
```diff
@@ -1,20 +1,55 @@
...
+        {
+            game = new Game();
+        }
+
+        /// <summary>
+        /// Method <c>CheckRange()</c> to check if the value of each die is valid.
+        /// </summary>
+        public void CheckRange()
+        {
+            Debug.Assert(game.GetDie1().CurrentValue >= 1 && game.GetDie1().CurrentValue <= 6, "Invalid Die Value");
```
Comment (Rarisma) [inline CMP1903_A1_2324/Testing.cs]: Could these be split up for a more comprehensive testing experience?
If you were to do it this way it should allow you to find out what exactly is wrong with a roll, i.e you would instantly know if it's too high or too low?

**Review by Rarisma — COMMENTED** [review]
Here's what I like about your implementation:

- Great use of encapsilation
- Good use of comments, including XML ones.
- Good use of defensive programming / Error handling.
- Correct use and implementation of testing
- Meets most if not all requirements

I am unsure that your code completely conforms to the google c# style guide, for example the brace style used in your code and default c# does not conform to the c# style from my understanding.

Other than this and the other comment I made regarding your testing class; I believe your code is excellent and is easy to read. It is clear you understand the objectives of the assignment and how to use github and other relevant IDE tooling.

File: CMP1903_A1_2324/Program.cs (hunk 1 above)
Comment (KirbyXIV) [inline CMP1903_A1_2324/Program.cs:43]: good error handling, further expands on this on line 55

File: CMP1903_A1_2324/Game.cs [hunk 6]
```diff
@@ -1,21 +1,93 @@
...
+    /// </summary>
     internal class Game
-    {
-        /*
-         * The Game class should create three die objects, roll them, sum and report the total of the three dice rolls.
-         *
-         * EXTRA: For extra requirements (these aren't required though), the dice rolls could be managed so that the
-         * rolls could be continous, and the totals and other statistics could be summarised for example.
-         */
+    {
```
Comment (KirbyXIV) [inline CMP1903_A1_2324/Game.cs:17]: great XML comments

**Comment by KirbyXIV** [comment]
Overall a good bit of code, shows you know what is asked for in the brief. good job!

**Review by Rarisma — APPROVED** [review]
Pr approved

**Review by KirbyXIV — APPROVED** [review]

**Pull request merged**